fixes #145 request a connection

SnehaParuchuri opened the pull request:
Created API to request a connection, please review

VishnuJammula commented on backend/models/ConnectionRequest.go:
```diff
@@ -0,0 +1,9 @@
+type ConnectionRequest struct {
+	RequestID   int   `gorm:"primary_key" json:"requestId"`
+	RequestedBy int   `json:"requestedBy"`
+	ConnectTo   int   `json:"connectTo"`
```
@SnehaParuchuri Just thinking how "RequestedTo" would sound like

VishnuJammula replied:
and also the JSON field

SnehaParuchuri replied:
I've named them thinking how it'd sound in a sentence - "Connection with ID is RequestedBy <user1> to ConnectTo <user2>" seemed better than "Connection with ID is RequestedBy <user1> to RequestTo <user2>"

VishnuJammula replied:
Let's keep it **From** and **To** for simplicity

VishnuJammula commented on backend/models/ConnectionRequest.go:
```diff
@@ -0,0 +1,9 @@
+	RequestedBy int   `json:"requestedBy"`
+	ConnectTo   int   `json:"connectTo"`
+	CreatedAt   int64 `json:"createdAt"`
+	UpdatedAt   int64 `json:"updatedAt"`
```
Remove this field unless there is a requirement to update a Pending request.

SnehaParuchuri replied:
I thought of removing the records from this table when the other user accepts request and updated filed can be removed. Instead if we were to join this table with users table to get the followers list it can be renamed as "acceptedOn" and a new column to check the "pending/accepted" state should be added.

VishnuJammula replied:
Let's avoid keeping pending connections, following data in a single table. As you said, accepting a connection request must remove that entry from ConnectionRequests table and update the userid array. Since ConnectionRequests table only holds pending connections, there is no need for "pending/accepted" field. So I think UpdatedAt field can be removed safely.

VishnuJammula commented on backend/models/User.go:
```diff
@@ -5,19 +5,19 @@ import (
+	Password       string        `gorm:"type:text" json:"password"`
+	DisplayPicture string        `gorm:"type:text" json:"picture"`
+	ResumePath     string        `gorm:"type:text" json:"resumepath"`
+	Following      pq.Int64Array `gorm:"type:integer[]" json:"following"`
```
Wondering instead of using a list of user following id's, Lets have a separate table to map user to connections. Lets think about how to retrieve list of following by joining the connections table and User table.connections table could look like

ConnectionId | ConnectionFrom | Connection To | Date Connected

Please do let me know your thoughts on it.

SnehaParuchuri replied:
In case of retrieving the followers from another table, the query would have to search through all the rows and get the followerIds and another filed can be added to check if the connection is accepted or pending, if it's retrieved along with user a simple append can be used to populate the select from query to retrieve posts (I think the latter case would be faster)

VishnuJammula replied:
Okay, lets keep array of user id's that the user is connected but make sure the connection is bidirectional, meaning both **from & to** following lists must be updated. Not sure how we can tell the date when the connection is accepted in case we need to display it in UI.

SnehaParuchuri replied:
I've left a comment in the Discussions. Will update the structure if UI team wants to display the date.

VishnuJammula reviewed (changes requested):
@SnehaParuchuri Please go through these changes and let me know if you have any queries.

SnehaParuchuri commented:
@VishnuJammula updated my view on the requested changes, let me know your final opinion on them.

SnehaParuchuri commented:
Updated fields in ConnectionRequest model. Renamed the changed fields in controller and added few lines to check existing connection, already connected users and stuff
